File: native/android/voice/sakha/WakeWordMatcher.kt
/**
 * WakeWordMatcher — pure phrase-matching helper for Sakha wake-word.
 *
 * Used by SakhaWakeWordDetector (Step 24) to decide whether a transcript
 * snippet from the always-on SpeechRecognizer loop contains one of the
 * configured wake phrases. Kept as a pure object with no Android
 * dependencies so it's JVM-unit-testable (Step 23).
 *
 * Match contract:
 *   - case-insensitive (Latin script — Devanagari has no case so this
 *     is a no-op for "हे सखा")
 *   - punctuation-tolerant: "Hey, Sakha!" matches "hey sakha", "हे सखा।"
 *     matches "हे सखा" (the Devanagari dandas U+0964 / U+0965 are
 *     treated as punctuation alongside ASCII !?.,;:—–•…)
 *   - whitespace-tolerant: "  HEY    sakha  " matches "hey sakha"
 *   - word-boundary aware: "sakha" matches "Hey, Sakha!" but NOT
 *     "Sasakhayan" — implemented by padding with spaces on both sides
 *     of the normalized haystack and the normalized phrase, so a
 *     successful contains() match implies the phrase begins and ends
 *     at a token boundary
 *
 * Returns the matched phrase (in normalized form) so the caller can
 * log it for telemetry without leaking the raw transcript. Returns
 * null when no phrase matches.
 *
 * Complexity: O(P × N) where P is the number of phrases and N is the
 * transcript length. Both are small in practice (≤ 6 phrases, ≤ 200
 * chars of transcript) so this comfortably fits the < 5ms scan budget
 * the always-on detector loop allows per partial result.
 */

package com.mindvibe.kiaan.voice.sakha

object WakeWordMatcher {

    // ASCII punctuation + the Devanagari danda (U+0964) and double-danda
    // (U+0965), plus a few common Unicode dashes / bullets that
    // SpeechRecognizer occasionally injects.
    private val PUNCTUATION_RE = Regex("[\\p{Punct}—–•…।॥]+")
    private val WHITESPACE_RE = Regex("\\s+")

    /**
     * Lowercase + strip punctuation + collapse whitespace. Idempotent —
     * normalize(normalize(x)) == normalize(x).
     */
    fun normalize(text: String): String {
        if (text.isEmpty()) return ""
        val lowered = text.lowercase()
        val unpunct = PUNCTUATION_RE.replace(lowered, " ")
        val collapsed = WHITESPACE_RE.replace(unpunct, " ")
        return collapsed.trim()
    }

    /**
     * Return the first phrase from [phrases] (in normalized form) that
     * appears as a complete word-aligned phrase in [transcript].
     * Returns null if none match.
     *
     * Phrases are tested in the order given, so the caller can rank
     * preferred phrases (e.g. multi-word "hey sakha" before single-word
     * "sakha") to disambiguate when both could match.
     */
    fun match(transcript: String, phrases: List<String>): String? {
        if (transcript.isBlank() || phrases.isEmpty()) return null
        // Pad both sides so contains() implies a word-boundary match.
        val haystack = " " + normalize(transcript) + " "
        for (phrase in phrases) {
            val needle = normalize(phrase)
            if (needle.isEmpty()) continue
            val padded = " $needle "
            if (haystack.contains(padded)) return needle
        }
        return null
    }
}

File: kiaanverse-mobile/apps/mobile/native/ios/Sources/Sakha/WakeWordMatcher.swift
/**
 * WakeWordMatcher (iOS port)
 *
 * Direct port of apps/mobile/native/android/.../WakeWordMatcher.kt. Used
 * by the (M4) wake-word detector to match wake phrases inside transcripts
 * from SFSpeechRecognizer's partial results.
 *
 * Match contract:
 *   - case-insensitive (Devanagari has no case so no-op for "हे सखा")
 *   - punctuation-tolerant (ASCII + Devanagari danda U+0964/U+0965)
 *   - whitespace-tolerant
 *   - word-boundary aware (haystack and needle padded with spaces so a
 *     successful contains() match implies a token boundary)
 *
 * Returns the matched (normalized) phrase, never the raw transcript.
 *
 * O(P × N) — P phrases, N transcript length. Both small in practice
 * (≤6 phrases, ≤200 chars). Comfortable under the 5ms budget the M4
 * always-on detector loop allows per partial result.
 */

import Foundation

public enum WakeWordMatcher {

    private static let punctuationRe = try! NSRegularExpression(
        pattern: "[\\p{Punct}—–•…।॥]+"
    )
    private static let whitespaceRe = try! NSRegularExpression(pattern: "\\s+")

    /// Lowercase + strip punctuation + collapse whitespace. Idempotent.
    public static func normalize(_ text: String) -> String {
        if text.isEmpty { return "" }
        let lowered = text.lowercased()
        let unpunct = punctuationRe.stringByReplacingMatches(
            in: lowered,
            range: NSRange(lowered.startIndex..., in: lowered),
            withTemplate: " "
        )
        let collapsed = whitespaceRe.stringByReplacingMatches(
            in: unpunct,
            range: NSRange(unpunct.startIndex..., in: unpunct),
            withTemplate: " "
        )
        return collapsed.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    /// Return the first phrase from `phrases` (in normalized form) that
    /// appears as a complete word-aligned phrase in `transcript`. Returns
    /// nil if none match. Phrases are tested in the order given.
    public static func match(transcript: String, phrases: [String]) -> String? {
        if transcript.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { return nil }
        if phrases.isEmpty { return nil }
        let haystack = " " + normalize(transcript) + " "
        for phrase in phrases {
            let needle = normalize(phrase)
            if needle.isEmpty { continue }
            let padded = " \(needle) "
            if haystack.contains(padded) { return needle }
        }
        return nil
    }
}
